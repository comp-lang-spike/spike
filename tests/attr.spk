#!/usr/bin/env spike


class Foo {
    obj attr;
    
    attr {
        return attr;
    }
    attr = newValue {
        attr = newValue;
    }
}


printAttr(f) {
    stdout.printf("The attribute value is %d.\n", f.attr);
}

printWholeExpr(tmp) {
    stdout.printf("The value of the whole expression is %d.\n", tmp);
}

printTmp(tmp) {
    stdout.printf("The temporary value is %d.\n", tmp);
}

printOldAttr(tmp) {
    stdout.printf("The attribute value was %d.\n", tmp);
}


main(argv) {
    obj f, tmp;
    
    f = Foo.new();
    tmp = (f.attr = 42);
    printAttr(f);
    printWholeExpr(tmp);
    f.attr += 2;
    printAttr(f);
    
    tmp = ++f.attr;
    printAttr(f);
    printTmp(tmp);
    tmp = f.attr++;
    printAttr(f);
    printOldAttr(tmp);
    
    
    stdout.printf("\n* Reprise *\n");
    
    obj sym;
    sym = $attr;
    
    tmp = ((f.*sym) = 42);
    printAttr(f);
    printWholeExpr(tmp);
    (f.*sym) += 2;
    printAttr(f);
    
    tmp = ++(f.*sym);
    printAttr(f);
    printTmp(tmp);
    tmp = (f.*sym)++;
    printAttr(f);
    printOldAttr(tmp);
    
    return 0;
}
