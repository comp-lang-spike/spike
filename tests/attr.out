The attribute value is 42.
The value of the whole expression is 42.
The attribute value is 44.
The attribute value is 45.
The temporary value is 45.
The attribute value is 46.
The attribute value was 45.

* Reprise *
The attribute value is 42.
The value of the whole expression is 42.
The attribute value is 44.
The attribute value is 45.
The temporary value is 45.
The attribute value is 46.
The attribute value was 45.
